File: 55Jump_Game.cpp
class Solution {
public:
    bool canJump(vector<int>& nums) {
        int pos=0;
        int n = nums.size();
        for(int i=0; i<n; i++){
            if(i<=pos){
                pos = ((i + nums[i]) > pos)? (i+nums[i]) : pos;
                if(pos>=n-1){
                    return true;
                }
            }
            
        }
        return false;
    }
};

File: 55.jump-game.py
#
# @lc app=leetcode id=55 lang=python3
#
# [55] Jump Game
#

# @lc code=start
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        # greedy solution
        # from the end, backward shifting the right edge
        goal = len(nums) - 1
        for i in range(len(nums)-1,-1,-1):
            if i + nums[i] >= goal:
                # shrink the right boundry to the current position
                goal = i
        return goal == 0
    
                
                
# @lc code=end

